File: QuranMac/Sources/QuranMac/Services/QuranService.swift
import Foundation

class QuranService {
    static let shared = QuranService()
    private let baseURL = "https://api.quran.com/api/v4"
    
    private init() {}
    
    func fetchChapters() async throws -> [Chapter] {
        let url = URL(string: "\(baseURL)/chapters?language=en")!
        let (data, _) = try await URLSession.shared.data(from: url)
        let response = try JSONDecoder().decode(ChaptersResponse.self, from: data)
        return response.chapters
    }
    
    func fetchVerses(chapterId: Int) async throws -> [Verse] {
        // Fetching all verses for a chapter. Using per_page=300 to likely get all in one go for most chapters.
        // ideally we should handle pagination, but for simplicity we start with a large limit.
        let url = URL(string: "\(baseURL)/verses/by_chapter/\(chapterId)?language=en&fields=text_uthmani&per_page=300")!
        let (data, _) = try await URLSession.shared.data(from: url)
        let response = try JSONDecoder().decode(VersesResponse.self, from: data)
        return response.verses
    }
    
    func fetchReciters() async throws -> [Reciter] {
        let url = URL(string: "\(baseURL)/resources/recitations?language=en")!
        let (data, _) = try await URLSession.shared.data(from: url)
        let response = try JSONDecoder().decode(RecitationsResponse.self, from: data)
        return response.recitations
    }
    
    func fetchChapterAudio(reciterId: Int, chapterId: Int) async throws -> String {
        let url = URL(string: "\(baseURL)/chapter_recitations/\(reciterId)/\(chapterId)")!
        let (data, _) = try await URLSession.shared.data(from: url)
        let response = try JSONDecoder().decode(ChapterRecitationResponse.self, from: data)
        return response.audioFile.audioUrl
    }
}

File: QuranWindows/Services/QuranService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json;
using QuranWindows.Models;

namespace QuranWindows.Services;

public class QuranService
{
    private static QuranService? _shared;
    public static QuranService Shared => _shared ??= new QuranService();

    private readonly HttpClient _httpClient;
    private const string BaseUrl = "https://api.quran.com/api/v4";

    private QuranService()
    {
        _httpClient = new HttpClient();
    }

    public async Task<List<Chapter>> FetchChaptersAsync()
    {
        var url = $"{BaseUrl}/chapters?language=en";
        var json = await _httpClient.GetStringAsync(url);
        var response = JsonConvert.DeserializeObject<ChaptersResponse>(json);
        return response?.Chapters ?? new List<Chapter>();
    }

    public async Task<List<Verse>> FetchVersesAsync(int chapterId)
    {
        // Using per_page=300 to match Swift implementation
        var url = $"{BaseUrl}/verses/by_chapter/{chapterId}?language=en&fields=text_uthmani&per_page=300";
        var json = await _httpClient.GetStringAsync(url);
        var response = JsonConvert.DeserializeObject<VersesResponse>(json);
        return response?.Verses ?? new List<Verse>();
    }

    public async Task<List<Reciter>> FetchRecitersAsync()
    {
        var url = $"{BaseUrl}/resources/recitations?language=en";
        var json = await _httpClient.GetStringAsync(url);
        var response = JsonConvert.DeserializeObject<RecitationsResponse>(json);
        return response?.Recitations ?? new List<Reciter>();
    }

    public async Task<string> FetchChapterAudioAsync(int reciterId, int chapterId)
    {
        var url = $"{BaseUrl}/chapter_recitations/{reciterId}/{chapterId}";
        var json = await _httpClient.GetStringAsync(url);
        var response = JsonConvert.DeserializeObject<ChapterRecitationResponse>(json);
        return response?.AudioFile?.AudioUrl ?? string.Empty;
    }
}
